File: front/src/features/play/utils/xp.ts
export const MAX_LEVEL = 99;
export const MAX_XP = 2_425_500;

export function xpForLevel(level: number): number {
	return (500 * level * (level - 1)) / 2;
}

export function levelFromXp(xp: number): number {
	if (xp <= 0) return 1;
	return Math.min(Math.floor((1 + Math.sqrt(1 + (8 * xp) / 500)) / 2), MAX_LEVEL);
}

export function xpForNextLevel(level: number): number {
	return 500 * level;
}

export function levelProgress(xp: number, level: number): number {
	if (level >= MAX_LEVEL) return 100;
	const currentXp = xp - xpForLevel(level);
	const needed = xpForNextLevel(level);
	if (needed <= 0) return 0;
	return Math.max(0, Math.min(100, (currentXp / needed) * 100));
}

export function exactLevel(xp: number): number {
	const level = levelFromXp(xp);
	return level + levelProgress(xp, level) / 100;
}

File: back/User-API/src/xp.rs
pub const MAX_LEVEL: i64 = 99;
pub const MAX_XP: i64 = 2_425_500;

pub fn xp_for_level(level: i64) -> i64 {
    500 * level * (level - 1) / 2
}

pub fn level_from_xp(xp: i64) -> i64 {
    if xp <= 0 {
        return 1;
    }
    let n = (1.0 + (1.0 + 8.0 * xp as f64 / 500.0).sqrt()) / 2.0;
    (n.floor() as i64).min(MAX_LEVEL)
}

pub fn xp_for_next_level(level: i64) -> i64 {
    500 * level
}

pub fn level_progress(xp: i64, level: i64) -> f64 {
    if level >= MAX_LEVEL {
        return 100.0;
    }
    let current_xp = xp - xp_for_level(level);
    let needed = xp_for_next_level(level);
    if needed <= 0 {
        return 0.0;
    }
    (current_xp as f64 / needed as f64 * 100.0).clamp(0.0, 100.0)
}
